tests/sentry/auth/test_helper.py

```diff
@@ -77,25 +120,147 @@ def test_associated_existing_member_invite_request(self):
         assert not OrganizationMember.objects.filter(id=member.id).exists()
 
     def test_associate_pending_invite(self):
-        provider = AuthProvider.objects.create(organization=self.organization, provider="dummy")
-        identity = {"id": "1234", "email": "test@example.com", "name": "Morty"}
-
         # The org member invite should have a non matching email, but the
         # member id and token will match from the cookie, allowing association
         member = OrganizationMember.objects.create(
             organization=self.organization, email="different.email@example.com", token="abc"
         )
 
-        request = RequestFactory().post("/auth/sso/")
-        request.user = AnonymousUser()
-        request.COOKIES["pending-invite"] = urlencode(
+        self.request.COOKIES["pending-invite"] = urlencode(
             {"memberId": member.id, "token": member.token, "url": ""}
         )
 
-        auth_identity = handle_new_user(provider, self.organization, request, identity)
+        auth_identity = self._handle_new_user()
 
         assigned_member = OrganizationMember.objects.get(
             organization=self.organization, user=auth_identity.user
         )
 
         assert assigned_member.id == member.id
+
+
+class HandleExistingIdentityTest(AuthIdentityHandlerTest):
+    @mock.patch("sentry.auth.helper.auth")
+    def test_simple(self, mock_auth):
+        mock_auth.get_login_redirect.return_value = "test_login_url"
+        user, auth_identity = self.set_up_user_identity()
+
+        redirect = handle_existing_identity(
+            self.auth_provider,
+            self.provider_obj,
+            self.organization,
+            self.request,
+            self.state,
+            auth_identity,
+            self.identity,
+        )
+
+        assert redirect.url == mock_auth.get_login_redirect.return_value
+        assert mock_auth.get_login_redirect.called_with(self.request)
+
+        persisted_identity = AuthIdentity.objects.get(ident=auth_identity.ident)
+        assert persisted_identity.data == self.identity["data"]
+
+        persisted_om = OrganizationMember.objects.get(user=user, organization=self.organization)
+        assert getattr(persisted_om.flags, "sso:linked")
+        assert not getattr(persisted_om.flags, "sso:invalid")
+
+        login_request, login_user = mock_auth.login.call_args.args
+        assert login_request == self.request
+        assert login_user == user
+
+
+class HandleAttachIdentityTest(AuthIdentityHandlerTest):
+    def _handle_attach_identity(self):
+        return handle_attach_identity(
+            self.auth_provider, self.request, self.organization, self.provider_obj, self.identity
+        )
+
+    @mock.patch("sentry.auth.helper.messages")
+    def test_new_identity(self, mock_messages):
```

Comment:
could this test also check that an OrganizationMember was created // should we make another case where an OM already exists?

Reply:
Yes to both.